### PR TITLE
Adding support for QNX OS

### PR DESCRIPTION
**Pull request by SebastianSchildt**
This enables tokio to be used on QNX.
It also depends on

https://github.com/tokio-rs/mio/pull/1766

@AkhilTThomas I added you to the fork, so you can write there as well

The changes are rather trivial, the main thing is a i32/u32 type mismatch that is handled

Not so sure what is expected/can done in terms of testing, as I assume there are no QNX machines in CI? What are the expectations here?

Once this is in acceptable form, would we need to backport it on one of the many version branches?
As just tokio users we would appreciate some form of QNX support being available out of the box on released tokio versions, (even if would be hidden behind a "here be dragons, we have no idea what this does" kind-of flag )

### CODE REVIEW

File: tokio/src/process/mod.rs
Comment (Darksonn) [inline tokio/src/process/mod.rs:678]:
```suggestion
        #[cfg(target_os = "nto")]
        let id = id as u32;
        self.std.uid(id);
```

**Comment by Darksonn** [comment]
Thanks for the PR. On the assumption that the mio PR is merged, then I am happy to add it here as well.

> Once this is in acceptable form, would we need to backport it on one of the many version branches?

No, it will just go in with the next minor release and be available starting from Tokio v1.37 (or later depending on timing). Unlike mio, we don't have multiple supported major versions.

> Not so sure what is expected/can done in terms of testing, as I assume there are no QNX machines in CI? What are the expectations here?

Adding support in CI is preferred but not required. Often, `cargo check` is possible even if there are no QNX machines in CI if we use cross-compilation.

> (even if would be hidden behind a "here be dragons, we have no idea what this does" kind-of flag )

Tokio has [a list of platforms with guaranteed support](https://docs.rs/tokio/latest/tokio/#supported-platforms) and a bunch of other platforms we support on a best-effort basis. This will not be something we _guarantee_ ongoing support for, but I do not mind adding it.

File: tokio/src/net/unix/ucred.rs
Comment (gh-tr) [inline tokio/src/net/unix/ucred.rs:164]: Looking at my diff for tokio 1.18.2, I had enabled the impl_netbsd path, not the impl_bsd path.  I presumably did this as it includes the PID whereas getpeereid() does not.  Any argument why the getpeereid() path should be used over this?

**Review by gh-tr — COMMENTED** [review]
Curious about the bsd versus netbsd path.

File: tokio/src/net/unix/ucred.rs
Comment (Darksonn) [reply tokio/src/net/unix/ucred.rs:164]: I don't know what is the correct answer for this target.

Comment (AkhilTThomas) [reply tokio/src/net/unix/ucred.rs:164]: You are right, using the impl_netbsd is better. I will update the PR

**Comment by SebastianSchildt** [comment]
Running test suite would need https://github.com/nix-rust/nix/pull/2055 not a blocker now for "best effort" QNX support, but preferably should be solved

**Comment by SteveLauC** [comment]
Hi, Nix maintainer here, that Nix PR wasn't merged because we think it [might not be worth considering how small the user group is](https://github.com/nix-rust/nix/pull/2055#discussion_r1285267681), but if tokio wants this target support and it is blocked by that Nix PR, then I am willing to merge it.

**Comment by Darksonn** [comment]
That's needed to compile the test-suite, but would it actually enable us to run the tests? Can we run this OS in CI?

**Comment by SebastianSchildt** [comment]
Hi @Darksonn : Yes, it would enable to compile, run tests. Which is good. Running all this in CI seems more complicated, see also my comment on mio https://github.com/tokio-rs/mio/pull/1766#issuecomment-2029037996

Again, as QNX users we appreciate everything that moves QNX support a bit closer to be on-par with other Tier 1 targets, but there is still some way to go. How far individual projects are willing to go to support this -for now- niche target is of course up to the communities. First step would be having some "not guaranteed to work/best effort code", that gives some amount of out-of -the-box functionality. Obviously, for all users it would be good to be able to run the test suite at least locally, _especially_ since CI testing might still be a slightly harder nut to crack.

**Comment by Darksonn** [comment]
This has been merged for mio. Any updates on the PR for Tokio?

**Comment by AkhilTThomas** [comment]
> This has been merged for mio. Any updates on the PR for Tokio?

Hi @Darksonn We are waiting for the merge of the nix fix and one on the signal-hook-registry as well.

**Comment by Darksonn** [comment]
Is there a link to the signal-hook-registry change?

**Comment by AkhilTThomas** [comment]
> Is there a link to the signal-hook-registry change?

Unfortunately not yet, its stuck with an internal clearance 😢. But its a one liner with a suppression for SA_RESTART being unavailable for nto target.

**Comment by SebastianSchildt** [comment]
> > Is there a link to the signal-hook-registry change?
> 
> Unfortunately not yet, its stuck with an internal clearance 😢. 

That is was, it isn't anymore https://github.com/vorner/signal-hook/pull/158

**Comment by Darksonn** [comment]
Any updates here? I'm okay with merging it before it goes into mio, since the change is so small. (But I do have a pending change request.)

**Comment by SebastianSchildt** [comment]
Hi, it is not forgotten. May take a few days before sitting at a suitable computer. Just to be sure, the change you are referring is the shorter/ more "elegant" way of doing the ```id as i32;```, right?

**Comment by Darksonn** [comment]
Okay, that's fine.

And yes, that's what I meant.

File: tokio/src/process/mod.rs
Comment (SebastianSchildt) [reply tokio/src/process/mod.rs]: Should be fixed in @AkhilTThomas  commit https://github.com/tokio-rs/tokio/pull/6421/commits/29cfcf827ce1964fcfac0b58fa8666c54f7c6917

Comment (Darksonn) [reply tokio/src/process/mod.rs]: Please fix it everywhere.

**Comment by Darksonn** [comment]
You probably need to merge in master for CI to pass.

**Comment by Darksonn** [comment]
It looks like your branch is messed up.

**Comment by AkhilTThomas** [comment]
> It looks like your branch is messed up.

My bad. Fixed it

File: tokio/src/process/mod.rs
Comment (Darksonn) [inline tokio/src/process/mod.rs:677]: What happens if `uid` is given an id that becomes negative when cast to i32?

**Review by Darksonn — COMMENTED** [review]
Overall LGTM, just one question about the integer cast.

**Review by Darksonn — APPROVED** [review]
Thank you.

**Comment by justinlooney** [comment]
**Updated finding — 2026-05-29**

Root cause is more specific than originally documented.

libc 0.2.186 defines LOCAL_PEEREID and SO_PEERCRED for
`target_env = "nto70"` / `"nto71"` (QNX 7.0/7.1) but has
no `nto80` branch. QNX 8.0 system headers do not contain
these constants at all — confirmed by grepping
/opt/qnx800/sdp2/target/qnx/usr/include/.

Tokio PR #6421 (merged May 2024, tokio 1.52.3 in use) is
necessary but not sufficient — the compilation still fails
because libc has no nto80 definitions.

Two upstream fix paths:
1. libc: add explicit `target_env = "nto80"` arm to the
   cfg_if block — blocked, constants absent from QNX 8.0 headers
2. tokio + socket2: gate the nto socket code on
   `target_env = "nto70"` / `"nto71"` only, not the broader
   `target_os = "nto"` — so QNX 8.0 builds cleanly by
   omitting the unavailable socket options

Path 2 appears to be the correct fix.

Happy to test any patch against QNX SDP 8.0 (qcc 12.2.0,
aarch64-unknown-nto-qnx800 target)

**Pull request merged**
